grout/models.py

```diff
@@ -52,43 +75,138 @@ def validate_schema(self, schema):
         jsonschema.Draft4Validator.check_schema(schema)
 
 
-class Record(GroutModel):
-    """Spatiotemporal records -- e.g. Loch Ness Monster sightings, crime events, etc."""
+class AbstractFlexibleRecord(GroutModel):
+    """
+    Base class for flexible records.
+    """
+    schema = models.ForeignKey('RecordSchema', on_delete=models.CASCADE)
+    data = JSONField()
+    archived = models.BooleanField(default=False)
+
+    class Meta(object):
+        abstract = True
+
+
+class DateTimeRange(models.Model):
+    """
+    Base class for a flexible record with date and time attributes.
+    """
     occurred_from = models.DateTimeField()
     occurred_to = models.DateTimeField()
 
+    class Meta(object):
+        abstract = True
+
+
+class PointFields(models.Model):
+    """
+    Base class for a record with point geometry.
+    """
+    class GeometryTypes(object):
+        POINT = 'point'
+        CHOICES = ((POINT, 'Point'),)
+
     geom = models.PointField(srid=settings.GROUT['SRID'])
```

Comment:
I like this idea a lot! The class inheritance that's currently implemented in this PR makes the fields nicely composable, but I think it's reached a point where there are enough base classes and mixins now that it would be confusing to read if I didn't already understand the data model.

In terms of data validation, were you thinking of continuing to use a `geometry_type` field (or equivalent) in order to determine how the `Cast` should behave? That sounds like a promising idea to me, and it suggests that we could much more easily expand geometry support in the future without having to introduce millions of mixins. I haven't used `GeometryField` or `Cast` before and the GeoDjango docs are extremely sparse, so I have a suspicion that there may be a serious problem lurking there, but it seems reasonable to spend some time experimenting with that approach by setting up tests, sketching out the API, and seeing if it works.

Reply:
Yep, that's exactly what I was thinking for validation. I haven't used this particular combination of features either, so it's definitely possible that there are some problems hiding somewhere, but yeah, I agree it seems worth a shot!